src/Microsoft.DotNet.Build.Tasks/DownloadFilesFromUrl.cs

```diff
@@ -0,0 +1,174 @@
+        /// The items to download.
+        /// Url and DestinationFile are required in the item's metadata, DestinationDir is optional.
+        /// </summary>
+        [Required]
+        public ITaskItem[] Items { get; set; }
+
+        /// <summary>
+        /// The destination directory for all the items to be downloaded.
+        /// This parameter can be overriden per item, if an item has DestinationDir in it's metadata, we will take that directory.
+        /// If both, the item's and this property are empty, the items will be downloaded to the working directory.
+        /// </summary>
+        public string DestinationDir { get; set; }
+
+        /// <summary>
+        /// The default is to fail on error, set this to true if you want to just warn.
+        /// </summary>
+        public bool TreatErrorsAsWarnings { get; set; }
+
+        /// <summary>
+        /// The list of files created. It is not guaranted that all the input items will be successfully downloaded
+        /// when TreatErrorsAsWarings is set to true.
+        /// </summary>
+        [Output]
+        public ITaskItem[] FilesCreated { get; set; }
+
+        public override bool Execute()
+        {
+            var filesCreated = new List<ITaskItem>();
+            foreach (var item in Items)
+            {
+                string downloadSource = item.GetMetadata("Url");
+                if (!Uri.IsWellFormedUriString(downloadSource, UriKind.Absolute))
+                {
+                    if (TreatErrorsAsWarnings)
+                    {
+                        Log.LogWarning($"Item {item.ItemSpec} Url is not a valid url.");
+                        continue;
+                    }
+                    else
+                    {
+                        return ExitWithError($"Item {item.ItemSpec} Url is not a valid url.");
+                    }
+                }
+
+                Uri downloadUri = new Uri(downloadSource);
+                string fileName = item.GetMetadata("DestinationFile");
+                if (string.IsNullOrWhiteSpace(fileName))
+                {
+                    // starts with file://
+                    if (downloadUri.IsFile)
+                    {
+                        fileName = Path.GetFileName(downloadUri.LocalPath);
+                    }
+                    else
+                    {
+                        while (downloadSource.EndsWith("/"))
+                        {
+                            downloadSource = downloadSource.Substring(0, downloadSource.Length - 1);
+                        }
+
+                        string[] urlContents = downloadSource.Split('/');
+
+                        // we set the filename to whatever is at the end of the url.
+                        // with this we ensure that if the metadata doesn't contain a file name, we always get one.
+                        fileName = urlContents[urlContents.Length - 1];
+
+                        if (string.IsNullOrEmpty(fileName))
```

Comment:
I would move this check one level up just to be sure that the file case also gets a filename that isn't empty.